gateway/mw_oauth2_auth.go

```diff
@@ -204,19 +214,19 @@ func newOAuth2ClientCredentialsConfig(OAuthSpec *UpstreamOAuth) oauth2clientcred
 	}
 }
 
-func (cache *upstreamOAuthCache) obtainToken(ctx context.Context, OAuthSpec *UpstreamOAuth) (string, error) {
+func (cache *upstreamOAuthClientCredentialsCache) obtainToken(ctx context.Context, OAuthSpec *UpstreamOAuth) (*oauth2.Token, error) {
 	cfg := newOAuth2ClientCredentialsConfig(OAuthSpec)
 
 	tokenSource := cfg.TokenSource(ctx)
 	oauthToken, err := tokenSource.Token()
 	if err != nil {
-		return "", err
+		return &oauth2.Token{}, err
 	}
 
-	return oauthToken.AccessToken, nil
+	return oauthToken, nil
 }
 
-func (cache *upstreamOAuthCache) setTokenInCache(cacheKey, token string) error {
+func setTokenInCache(cacheKey string, token string, ttl time.Duration, cache *storage.RedisCluster) error {
```

Comment:
do we need to keep a cacher as a dependency in `upstreamOAuthClientCredentialsCache`, so that it can be reused when password flow is implemented?

Reply:
yes, I rewrote this with PasswordAuth in mind so that's why I'm passing the cacher here